Add option to stream logs to file for flutter logs and way to use it in devicelab runs

dnfield opened the pull request:
Fixes https://github.com/flutter/flutter/issues/84000

See also https://github.com/flutter/flutter/issues/83939, https://github.com/flutter/flutter/issues/77193

We'll need a recipe change to get this running on the recipes.

jonahwilliams commented on packages/flutter_tools/test/commands.shard/hermetic/logs_test.dart:
```diff
@@ -22,11 +32,80 @@ void main() {
+
+  @override
+  String toString() => 'FakeLogReader';
+}
```
```suggestion
}

```

jonahwilliams reviewed (approved):
tool changes LGTM

jmagman commented:
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8845332628763500560/+/u/run_test.dart_for_framework_tests_shard_and_subshard_misc/stdout
```
  Unhandled exception:
  Invalid argument(s): Could not find an option named "device-logs-path".
  #0      ArgResults.[] (package:args/src/arg_results.dart:65:7)
  #1      main (file:///b/s/w/ir/x/w/flutter/dev/devicelab/bin/run.dart:104:35)
  #2      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:281:32)
  #3      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

jmagman commented on dev/devicelab/bin/run.dart:
```diff
@@ -101,6 +101,13 @@ Future<void> main(List<String> rawArgs) async {
     return;
   }
 
+  final String/*?*/ logPath = args['device-logs-path'] as String/*?*/;
```
The arg is `device-log-path`.  Probably worth adding `device-logs-path` as an alias to the arg if you, the person who wrote it a few minutes ago, couldn't quite remember which it was 😄

dnfield replied:
Yes...

dnfield replied:
WDYT of doing this as an optional arg vs. detecting `LocalPlatform().environment[FLUTTER_LOGS_DIR]`?

I mentionedin the issue, I'd rather avoid the `HostAgent` logic since I don't want to do this by default on local dev machines.

dnfield replied:
As discussed in chat, uploaded a new change to make hostAGent not fall back to a local temp dir (and updated the test to make sure that's the case), and now using hostAgent.

jmagman commented on dev/devicelab/lib/framework/host_agent.dart:
```diff
@@ -20,22 +20,19 @@ class HostAgent {
 
   /// Creates a directory to dump file artifacts.
-  Directory get dumpDirectory {
+  Directory/*?*/ get dumpDirectory {
```
Thanks for doing this.

jmagman commented on packages/flutter_tools/test/commands.shard/hermetic/logs_test.dart:
```diff
@@ -22,11 +32,80 @@ void main() {
+      _FakeDevice(),
+    ];
+  }
+
```
```suggestion
```

jmagman reviewed (approved):
LGTM

dnfield committed:
renmae file

dnfield committed:
rename file

dnfield committed:
working

dnfield commented:
Ok, now this works.

I've renamed `adb.dart` to `devices.dart` since it contains .. everything about devices and little about adb.

I've added new methods to `Device` to start/stop logging with minimal filtering (--quiet on iOS, a bit more restrictive on Android but keeping all fatal logs).

I've wrapped task execution with actually collecting those logs.

dnfield commented:
Most of the changes here are simple file renames. https://github.com/flutter/flutter/pull/84008/commits/a3bfe1e127ac26cceb8f8cf4bff8c02d05941d92 is the meat of this change.

dnfield committed:
work even if no device found

dnfield committed:
merge

dnfield commented:
Fixed up the merge conflicts.

jmagman reviewed (approved):
LGTM
Thanks for renaming `adb.dart`, that should have been done awhile ago!

chinmaygarde commented:
Since Dan is OOO this week, can we land this right away please?

dnfield commented:
Oh I landed it. Not waiting ror another merge conflict. I'm semi-around.
